provider/core/src/versions.rs

```diff
@@ -0,0 +1,82 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[derive(Debug, Clone, Default, Copy, PartialEq)]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[non_exhaustive]
+/// Source data versions
+pub struct SourceVersions<'data> {
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    /// The version of CLDR source data, if used during datagen.
+    pub cldr: Option<SourceVersion<'data>>,
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    /// The version of ICU source data, if used during datagen.
+    pub icuexport: Option<SourceVersion<'data>>,
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    /// The version of segmenter LSTM data, if used during datagen.
+    pub segmenter_lstm: Option<SourceVersion<'data>>,
+}
+
+impl SourceVersions<'_> {
+    /// Creates an empty struct.
+    pub const fn empty() -> Self {
+        SourceVersions {
+            cldr: None,
+            icuexport: None,
+            segmenter_lstm: None,
+        }
+    }
+}
+
+#[cfg(feature = "datagen")]
+impl databake::Bake for SourceVersions<'_> {
+    fn bake(&self, env: &databake::CrateEnv) -> databake::TokenStream {
+        env.insert("icu_provider");
+
+        if self == &Self::empty() {
+            return databake::quote!("icu_provider::SourceVersions::empty()")
+        }
+
+        let mut setters = databake::quote!();
+        if let Some(cldr) = &self.cldr {
+            let cldr = cldr.bake(env);
+            setters = databake::quote! {
+                #setters
+                v.cldr = Some(#cldr);
+            }
+        };
+        if let Some(icuexport) = &self.icuexport {
+            let icuexport = icuexport.bake(env);
+            setters = databake::quote! {
+                #setters
+                v.icuexport = Some(#icuexport);
+            }
+        };
+        if let Some(segmenter_lstm) = &self.segmenter_lstm {
+            let segmenter_lstm = segmenter_lstm.bake(env);
+            setters = databake::quote! {
+                #setters
+                v.segmenter_lstm = Some(#segmenter_lstm);
+            }
+        };
+        databake::quote! {{
+            let mut v = icu_provider::SourceVersions::empty();
+            #setters
+            v
+        }}
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[non_exhaustive]
+#[cfg_attr(feature = "datagen", derive(databake::Bake, serde::Serialize), databake(path = icu_provider))]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+/// A concrete version
+pub enum SourceVersion<'data> {
+    /// A released version
+    Public(#[cfg_attr(feature = "serde", serde(borrow))] &'data str),
```

Comment:
Suggestion: use a Cow so we don't leak strings.

Reply:
Can't have cows in consts because of dropping stuff.